langchain4j-infinispan/src/main/java/dev/langchain4j/store/embedding/infinispan/InfinispanEmbeddingStore.java

```diff
@@ -0,0 +1,249 @@
+                                    String name,
+                                    Integer dimension) {
+        ensureNotNull(builder, "builder");
+        ensureNotBlank(name, "name");
+        ensureNotNull(dimension, "dimension");
+        itemMarshaller = new LanchainItemMarshaller(dimension);
+        builder.remoteCache(name)
+              .configuration(DEFAULT_CACHE_CONFIG.replace("CACHE_NAME", name)
+                    .replace("LANCHAINITEM", itemMarshaller.getTypeName()));
+
+        // Registers the schema on the client
+        ProtoStreamMarshaller marshaller = new ProtoStreamMarshaller();
+        SerializationContext serializationContext = marshaller.getSerializationContext();
+        FileDescriptorSource fileDescriptorSource = new FileDescriptorSource();
+        String fileName = "lanchain_dimension_" + dimension + ".proto";
+        String fileContent = PROTO.replace("DIMENSION", dimension.toString());
+        fileDescriptorSource.addProtoFile(fileName, fileContent);
+        serializationContext.registerProtoFiles(fileDescriptorSource);
+        serializationContext.registerMarshaller(itemMarshaller);
+        builder.marshaller(marshaller);
+        // Uploads the schema to the server
+        RemoteCacheManager rmc = new RemoteCacheManager(builder.build());
+        RemoteCache<String, String> metadataCache = rmc
+              .getCache(ProtobufMetadataManagerConstants.PROTOBUF_METADATA_CACHE_NAME);
+        metadataCache.put(fileName, fileContent);
+
+        this.remoteCache = rmc.getCache(name);
+    }
+
+    @Override
+    public String add(Embedding embedding) {
+        String id = randomUUID();
+        add(id, embedding);
+        return id;
+    }
+
+    @Override
+    public void add(String id, Embedding embedding) {
+        addInternal(id, embedding, null);
+    }
+
+    @Override
+    public String add(Embedding embedding, TextSegment textSegment) {
+        String id = randomUUID();
+        addInternal(id, embedding, textSegment);
+        return id;
+    }
+
+    @Override
+    public List<String> addAll(List<Embedding> embeddings) {
+        List<String> ids = embeddings.stream()
+                .map(ignored -> randomUUID())
+                .collect(toList());
+        addAllInternal(ids, embeddings, null);
+        return ids;
+    }
+
+    @Override
+    public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedded) {
+        List<String> ids = embeddings.stream()
+                .map(ignored -> randomUUID())
+                .collect(toList());
+        addAllInternal(ids, embeddings, embedded);
+        return ids;
+    }
+
+    @Override
+    public List<EmbeddingMatch<TextSegment>> findRelevant(Embedding referenceEmbedding, int maxResults, double minScore) {
+        Query<Object[]> query = remoteCache.query("select i, score(i) from " + itemMarshaller.getTypeName() + " i where i.floatVector <-> "+ Arrays.toString(referenceEmbedding.vector()) +"~3");
```

Comment:
I don't know the Langchain4J use case. Do we always expect the 3-nearest values? If not you can parametrize it.

Reply:
@fax4ever this is the script  in Elastic .script(s -> s.inline(InlineScript.of(i -> i
                        // The script adds 1.0 to the cosine similarity to prevent the score from being negative.
                        // divided by 2 to keep score in the range [0, 1]
                        .source("(cosineSimilarity(params.query_vector, 'vector') + 1.0) / 2")